go-controller/pkg/node/udn_isolation.go

```diff
@@ -0,0 +1,459 @@
+						} else {
+							waitingForActive = false
+						}
+					}
+				}
+			case err := <-errChan:
+				klog.Errorf("Systemd listener error: %v", err)
+			}
+		}
+	}()
+	return nil
+}
+
+func (m *UDNHostIsolationManager) podInitialSync() error {
+	pods, err := m.podLister.List(labels.Everything())
+	if err != nil {
+		return fmt.Errorf("failed to list pods: %v", err)
+	}
+	v4PodsIPs := map[string]sets.Set[string]{}
+	v6PodsIPs := map[string]sets.Set[string]{}
+	for _, pod := range pods {
+		// only add pods with primary UDN
+		primaryUDN, err := m.isPodPrimaryUDN(pod)
+		if err != nil {
+			return fmt.Errorf("failed to check if pod %s in namespace %s is in primary UDN: %w", pod.Name, pod.Namespace, err)
+		}
+		if !primaryUDN {
+			continue
+		}
+
+		podIPs, err := util.DefaultNetworkPodIPs(pod)
+		if err != nil {
+			return fmt.Errorf("failed to get pod IPs for pod %s in namespace %s: %v", pod.Name, pod.Namespace, err)
+		}
+		if podIPs == nil {
+			continue
+		}
+		newV4IPs, newV6IPs := splitIPsPerFamily(podIPs)
+		podKey, err := cache.MetaNamespaceKeyFunc(pod)
+		if err != nil {
+			klog.Warningf("UDNHostIsolationManager failed to get key for pod %s in namespace %s: %v", pod.Name, pod.Namespace, err)
+			continue
+		}
+		v4PodsIPs[podKey] = newV4IPs
+		v6PodsIPs[podKey] = newV6IPs
+	}
+	if err = m.udnPodIPsv4.fullSync(m.nft, v4PodsIPs); err != nil {
+		return err
+	}
+	if err = m.udnPodIPsv6.fullSync(m.nft, v6PodsIPs); err != nil {
+		return err
+	}
+	return nil
+}
+
+func podNeedsUpdate(oldObj, newObj *v1.Pod) bool {
+	if oldObj == nil || newObj == nil {
+		return true
+	}
+	// react to pod IP changes
+	return !reflect.DeepEqual(oldObj.Status, newObj.Status) || !reflect.DeepEqual(oldObj.Annotations, newObj.Annotations)
+}
+
+func (m *UDNHostIsolationManager) reconcilePod(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		klog.Errorf("UDNHostIsolationManager failed to split meta namespace cache key %s for pod: %v", key, err)
+		return nil
+	}
+	pod, err := m.podLister.Pods(namespace).Get(name)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			// Pod was deleted, clean up.
+			return m.updateUDNPodIPs(key, nil)
+		}
+		return fmt.Errorf("failed to fetch pod %s in namespace %s", name, namespace)
+	}
+	// only add pods with primary UDN
+	primaryUDN, err := m.isPodPrimaryUDN(pod)
+	if err != nil {
+		return fmt.Errorf("failed to check if pod %s in namespace %s is in primary UDN: %w", name, namespace, err)
+	}
+	if !primaryUDN {
+		return nil
+	}
+	podIPs, err := util.DefaultNetworkPodIPs(pod)
+	if err != nil {
+		// update event should come later with ips
+		klog.V(5).Infof("Failed to get default network pod IPs for pod %s in namespace %s: %v", name, namespace, err)
+		return nil
+	}
+	return m.updateUDNPodIPs(key, podIPs)
+}
+
+func (m *UDNHostIsolationManager) isPodPrimaryUDN(pod *v1.Pod) (bool, error) {
+	activeNetwork, err := m.nadController.GetActiveNetworkForNamespace(pod.Namespace)
```

Comment:
since you have the pod, you could look at the annotations to avoid having to use the nad controller for network with role: primary. I think there is already a util method somewhere to parse this for you.

Reply:
how do I know from the annotation if pod's primary network is UDN? Can you help me find that util function because I can't? :(

Reply:
```
    k8s.ovn.org/pod-networks: '{"default":{"ip_addresses":["10.244.5.4/24"],"mac_address":"0a:58:0a:f4:05:04","routes":[{"dest":"10.244.0.0/16","nextHop":"10.244.5.1"},{"dest":"100.64.0.0/16","nextHop":"10.244.5.1"}],"ip_address":"10.244.5.4/24","role":"infrastructure-locked"},"test-namespace-9/user-defined-network-9":{"ip_addresses":["10.20.7.3/24"],"mac_address":"0a:58:0a:14:07:03","gateway_ips":["10.20.7.1"],"routes":[{"dest":"10.20.0.0/16","nextHop":"10.20.7.1"},{"dest":"10.96.0.0/16","nextHop":"10.20.7.1"},{"dest":"100.65.0.0/16","nextHop":"10.20.7.1"}],"ip_address":"10.20.7.3/24","gateway_ip":"10.20.7.1","role":"primary"}}'
```

you see the role there. That tells you if it is a UDN or not:

```
	// defined internally by ovnkube to recognize "default"
	// network's role as a "infrastructure-locked" network
	// when user defined network is the primary network for
	// the pod which makes "default" network niether primary
	// nor secondary
	NetworkRoleInfrastructure = "infrastructure-locked"
```

You can unmarshal the pod annotations for the default network and then check role:
```
podAnnotation, err := util.UnmarshalPodAnnotation(pod.Annotations, types.DefaultNetworkName)
if annotation.Role ==NetworkRoleInfrastructure { return true
} else {
activeNetwork, err := m.nadController.GetActiveNetworkForNamespace(pod.Namespace)
}

Reply:
Oh I thought using `GetActiveNetworkForNamespace` is a new more efficient way of doing that because it doesn't require unmarshalling? Are you saying if you have pod object, you should prefer unmarshalling network annotations to get network info?

Reply:
UnmarshalPodAnnotation seems to be more reliable, as it dosn't depend on NAD presence (which can actually be deleted). I have skipped `GetActiveNetworkForNamespace` fallback and just returned error as no annotations means I can't extract ips anyways